pdns/svc-records.cc

```diff
@@ -0,0 +1,181 @@
+ * it under the terms of version 2 of the GNU General Public License as
+ * published by the Free Software Foundation.
+ *
+ * In addition, for the avoidance of any doubt, permission is granted to
+ * link this program with OpenSSL and to (re)distribute the binaries
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#include "svc-records.hh"
+#include "misc.hh"
+#include "base64.hh"
+
+const std::map<std::string, SvcParam::SvcParamKey> SvcParam::SvcParams = {
+  {"mandatory", SvcParam::SvcParamKey::mandatory},
+  {"alpn", SvcParam::SvcParamKey::alpn},
+  {"no-default-alpn", SvcParam::SvcParamKey::no_default_alpn},
+  {"port", SvcParam::SvcParamKey::port},
+  {"ipv4hint", SvcParam::SvcParamKey::ipv4hint},
+  {"echconfig", SvcParam::SvcParamKey::echconfig},
+  {"ipv6hint", SvcParam::SvcParamKey::ipv6hint}
+};
+
+SvcParam::SvcParamKey SvcParam::keyFromString(const std::string& k) {
+  auto it = SvcParams.find(k);
+  if (it != SvcParams.end()) {
+    return it->second;
+  }
+  if (k.substr(0, 3) == "key") {
+    try {
+      return SvcParam::SvcParamKey(pdns_stou(k.substr(3)));
+    }
+    catch (...) {
+    }
+  }
+  throw std::invalid_argument("SvcParam '" + k + "'is not recognized or in keyNNNN format");
+}
+
+std::string SvcParam::keyToString(const SvcParam::SvcParamKey& k) {
+  auto ret = std::find_if(SvcParams.begin(), SvcParams.end(), [&](const std::pair<std::string, SvcParam::SvcParamKey>& e) { return e.second == k; });
+  if (ret != SvcParams.end()) {
+    return ret->first;
+  }
+  return "key" + std::to_string(k);
+}
+
+SvcParam::SvcParam() {};
+
+SvcParam::SvcParam(const SvcParamKey &key) {
+  d_key = key;
+  if (d_key != SvcParamKey::no_default_alpn) {
+    throw std::invalid_argument("can not create non-empty SvcParam for key '" + keyToString(key) + "'");
+  }
+}
+
+SvcParam::SvcParam(const SvcParamKey &key, const std::string &value) {
+  d_key = key;
+  if (d_key != SvcParamKey::echconfig && d_key < 7) {
+    throw std::invalid_argument("can not create SvcParam for " + keyToString(key) + " with a string value");
+  }
+  if (d_key == SvcParamKey::echconfig) {
+    std::string d;
+    // TODO check Base64 decode
+    d_echconfig = value;
+    return;
+  }
+  d_value = value;
+}
+
+SvcParam::SvcParam(const SvcParamKey &key, const std::vector<std::string> & value) {
+  d_key = key;
+  if (d_key != SvcParamKey::alpn) {
+    throw std::invalid_argument("can not create SvcParam for " + keyToString(key) + " with a string-set value");
+  }
+  if (d_key == SvcParamKey::alpn) {
```

Comment:
this if looks redundant :)

Reply:
for now, yes. But there might be other string-set Service params in the future.